### ISSUE TITLE
Scaling of custom loss function

### ISSUE THREAD

**Issue by ytsaig** [dmlc/xgboost#1066]
When working with custom loss function through the python API, I am finding that xgboost is sensitive to the scaling of the gradient/hessian, which I would think should make no difference at all. 

Here's an example implementing squared loss with a constant scalar applied to the boston housing data.

```
import xgboost as xgb
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.datasets import load_boston
from sklearn.cross_validation import train_test_split
from sklearn.metrics import mean_squared_error

def squared_loss(preds, dtrain):
    labels = dtrain.get_label()
    grad = preds - labels
    hess = np.ones_like(grad)
    return grad, hess

def squared_loss_norm(preds, dtrain):
    labels = dtrain.get_label()
    grad = (preds - labels) / 1000.0
    hess = np.ones_like(grad) / 1000.0
    return grad, hess

boston = load_boston()
X_train, X_test, y_train, y_test = train_test_split(boston.data, np.log(boston.target), test_size=0.2, random_state=1)
feature_names = boston.feature_names

# Create DMatrix objects for xgboost
dtrain = xgb.DMatrix(X_train, y_train, feature_names=feature_names)
dtest = xgb.DMatrix(X_test, y_test, feature_names=feature_names)

# xgboost configuration
n_estimators = 100

# Train models
xgb_params = {'max_depth' : 3, 'eta' : 0.01, 'seed' : 12345, 'silent' : 1}
model_squared = xgb.train(xgb_params, dtrain, n_estimators, obj=squared_loss)
mse_squared = [mean_squared_error(y_test, model_squared.predict(dtest, ntree_limit=nt)) for nt in range(1, n_estimators+1)]

xgb_params = {'max_depth' : 3, 'eta' : 0.01, 'seed' : 12345, 'silent' : 1}
model_squared_norm = xgb.train(xgb_params, dtrain, n_estimators, obj=squared_loss_norm)
mse_squared_norm = [mean_squared_error(y_test, model_squared_norm.predict(dtest, ntree_limit=nt)) for nt in range(1, n_estimators+1)]

# Evaluate
mse = pd.DataFrame({'squared_loss' : mse_squared, 'squared_loss_norm' : mse_squared_norm}, index=range(1, n_estimators+1))
mse.plot()
plt.xlabel('Iterations')
plt.ylabel('MSE on test set')
```
This yields the following result:

![xgboost](https://cloud.githubusercontent.com/assets/1490131/14144153/475414f8-f697-11e5-81d2-af52f805956a.png)

So the learner makes no progress at all when fed the scaled grad/hessian, but works fine with the unscaled version. Am I missing something in the parameter settings?

**Comment by tqchen** [comment]
The parameter ```parameter min_child_weight``` and ```gamma``` all depends on scaling. As regularizer is a tradeoff between known information and prior. So if you have larger weights on your data, it is going to trust your data more  likely

See also http://xgboost.readthedocs.org/en/latest/model.html

**Comment by ytsaig** [comment]
Thanks, you're absolutely right. In this particular case, setting
`xgb_params = {'min_child_weight' : 1.0, 'lambda' : 0, 'max_depth' : 3, 'eta' : 0.1, 'seed' : 12345, 'silent' : 1}`
Leads to identical results between the scaled and the unscaled versions.
